coordinate-systems.md

```diff
@@ -0,0 +1,246 @@
+
+Currently, XRFrameOfReference refers to an enumerated set of standard coordinate systems relative to which all others must be able to be expressed. These include:
+
+
+
+*   _eye-level_: the origin and rotation of this frame of reference are the orientation of the device at the start of session.
+*   _stage_: the origin and rotation of this frame of reference are in the center of a known safe playing area like a VR room.
+*   _head-model_: this frame of reference is like eye-level except the translation component of the device pose is basically identity with a neck model. This is for experiences which expect the user to not move and only look around.
+
+These are basically different ways of expressing the transform of the device in 3D space. For example, virtual objects would have the same transform in _eye-level_, _stage_, and _head-model_ FORs, it's only the device that behaves differently (more on this later). _Stage_ basically allows for a virtual translation to be applied to the native device transform and _head-model_ always locks the device translation to (0, 0, 0).
+
+Finally, XRCoordinateSystem was not designed to be dynamic. If we want to use XRCoordinateSystems to track moving objects or reference frames, we need to design a system/API to handle that dynamicism. If we want to be able to express things like hit-test input rays relative to a device or input-source, we need to be able to represent that relationship. We also need to know what coordinate system the results are in.
+
+
+# Analysis
+
+
+## XRFrameOfReference
+
+
+### Basics
+
+The current API for using XRFrameOfReference (FOR) is as follows:
+
+
+```
+const frameOfReference = await session.requestFrameOfReference('stage');
+```
+
+
+Now that we have the frame of reference we want, we convert any XRCoordinateSystem into 'stage-space' by transforming by that FOR. Most simply, if we want the current transform of the device, we say:
+
+
+```
+let devicePose = xrFrame.getDevicePose(frameOfReference);
+```
+
+
+This is the pose of the device in the given FOR. If you want to think in terms of 'absolute space', the origin of all 3 styles of FORs is the start-of-session pose of the device - that is the device pose when the AR system was first activated. In fact, the documentation around the method XRFrameOfReference::TransformBasePose(TransformationMatrix base_pose) explains this well: "Transforms a given pose from a "base" coordinate system used by the XR service to the frame of reference's coordinate system." What this means is pretty simple - the conceit is that most transforms will naturally be expressed in some native system that has its own reference frame - in most cases it is relative to the start-of-service of the underlying AR/VR runtime.
```

Comment:
I'm uncomfortable with anything being defined relative to the initial pose of the device with WebXR is initialized, which makes a huge amount of the description here confusing (to me).

The initial pose of the device is meaningless. It could be sitting on the floor waiting to be picked up, on the top of my head, in my lap, in my hand pointing down.

Similarly, the coordinate system of the underlying platform is meaningless.  In current room scale VR systems, there is a "fixed" coordinate system.  But other systems can change their underlying base/device coordinate system over time:  the only meaningful relationships are between the reported device poses (head, controllers, etc) and any "anchors" or "reference frames" the platform supports (and those should be queried each frame).

We could assume that an implementation will drop an anchor when initialized, to remember what the starting coordinate frame is, but that's fragile.  If I drop and anchor and then walk a mile away wearing my HMD or carrying my phone, I have a hard time imagining we want to use that initial origin for anything, and an even harder time imagining it is still a useful value

Reply:
I think I may have failed to explain what I meant appropriately. By 'absolute space' i simply meant that there is some bedrock, native origin/coordinate system that the runtime generates for/as-part-of the device. Stage, which I think could be generalized to the kind of understanding Blair is talking about, is just some transform applied to the native layer's understanding of space that gets us into a more simplified or transformed system ('stage', 'head-model', or any other spatial reference that is useful to the app).